### ISSUE TITLE
ur'strings' are a Syntax Errors in Python 3

### ISSUE THREAD

**Issue by cclauss** [ckan/ckan#4039]
In Python 3, you can have __u'strings'__ and you can have __r'strings'__ but you can not have __ur'strings'__.
* __python3 -c "print(ur'string')"__  # --> Syntax Error

ur'strings' currently appear in three files...

$ __python3 -m flake8 . --count --select=E901,E999 --show-source --statistics__
```
./ckan/tests/test_coding_standards.py:127:55: E999 SyntaxError: invalid syntax
    pattern = re.compile(ur'#.*?coding[:=][ \t]*utf-?8')
                                                      ^
./ckan/lib/munge.py:159:42: E999 SyntaxError: invalid syntax
    filename = re.sub(ur'[^a-zA-Z0-9_. -]', '', filename).replace(u' ', u'-')
                                         ^
./doc/conf.py:341:49: E999 SyntaxError: invalid syntax
  ('contents', 'CKAN.tex', ur'CKAN documentation',
                                                ^
3     E999 SyntaxError: invalid syntax
```

The last one is governed by:
* http://www.sphinx-doc.org/en/stable/config.html#confval-latex_documents

Based on the [example __conf.py__ file](http://www.sphinx-doc.org/en/stable/config.html#example-of-configuration-file), I believe that we should use u'strings' in this context.  #4049

**Comment by sirex** [comment]
Related to https://github.com/ckan/ckan/issues/3309.

**Comment by cclauss** [comment]
Discussed in https://github.com/ckan/ckan/pull/4035#issuecomment-368993517

**Comment by cclauss** [comment]
@torfsen @amercader What about converting __ur'string'__ to __six.u(r'string')__?

```
python2 -c "print(ur'你好 Lüsai' == '你好 Lüsai')"    # False
python2 -c "print(ur'你好 Lüsai' == r'你好 Lüsai')"   # False
python2 -c "print(ur'你好 Lüsai' == u'你好 Lüsai')"   # True

python2 -c "import six ; print(ur'你好 Lüsai' == six.u('你好 Lüsai'))"     # True
python2 -c "import six ; print(ur'你好 Lüsai' == six.u(r'你好 Lüsai'))"    # True
# python2 -c "import six ; print(ur'你好 Lüsai' == six.u(u'你好 Lüsai'))"  # TypeError
```

**Comment by torfsen** [comment]
In general, we don't have many `ur` strings in CKAN:

```
$ grep -rI "\<ur[\"']" --include "*.py"
ckan/lib/munge.py:    filename = re.sub(ur'[^a-zA-Z0-9_. -]', '', filename).replace(u' ', u'-')
ckan/lib/munge.py:    filename = re.sub(ur'-+', u'-', filename)
ckan/tests/test_coding_standards.py:    pattern = re.compile(ur'#.*?coding[:=][ \t]*utf-?8')
ckan/tests/test_coding_standards.py:                        if (i > 1) and (line[i - 2:i].lower() == u'ur'):
doc/conf.py:  ('contents', 'CKAN.tex', ur'CKAN documentation',
doc/conf.py:   ur'CKAN contributors', 'manual'),
```

All of these are ASCII-only strings, so from an encoding point of view, nothing is lost when they are simply converted to `r` strings.

Judging from my experiments, the return type of Python 2's regular expression functions depends on the type of the input string, but not on the type of the pattern string:

```
>>> re.match(r'\w*', 'aä').group()
'a'
>>> re.match(r'\w*', u'aä').group()
u'a'
>>> re.match(ur'\w*', 'aä').group()
'a'
>>> re.match(ur'\w*', u'aä').group()
u'a'
```

Also, the type of the pattern string doesn't influence whether character classes like `\w` capture non-ASCII characters (as seen above, the `ä` is never matched, even when both the pattern and the input are `unicode`) -- that is decided by the presence of the `re.UNICODE` (or `re.U`) flag (in Python 2 at least).

In conclusion, it should be OK to convert all the `ur` instances of the current code base to `r` strings.

As mentioned before, we also should update [our Unicode handling documentation](https://github.com/ckan/ckan/blob/master/doc/contributing/unicode.rst) which currently says

> In CKAN, every string literal must carry either a u or a b prefix. While the latter is redundant in Python 2, it makes the developer's intention explicit and eases a future migration to Python 3.
>
> This rule also holds for raw strings, which are created using an r prefix. Simply use ur instead:
> 
>     m = re.match(ur'A\s+Unicode\s+pattern')

**Comment by sirex** [comment]
IMHO best option is to do:

```python
from __future__ import unicode_literals
```

Then there will be no need for `u''` strings at all and this will improved overall transition to Python 3.

**Comment by cclauss** [comment]
See last sentence of this section
* https://github.com/ckan/ckan/blob/master/doc/contributing/unicode.rst#string-literals

**Comment by sirex** [comment]
@cclauss thanks for pointnig that out, but as we see, that approach does not work with `ur''` strgins. And if according to documentation every unicode string has `u` prefix, then `unicode_literals` won't do any harm, bucause it supports `u` strgins too.

**Comment by smotornyuk** [comment]
they are converting any string into Unicode implicitly. Personally, i'm strongly against using it. Whereas it provides quite a nice way to move between versions, it hides potential bugs, so it would be better to run replace-regexp over code(or use one of existing converters 2to3).
if [this](http://python-future.org/unicode_literals.html) can be considered as an argument, I think that drawbacks are much more important in our case

**Comment by wardi** [comment]
I'm also strongly against unicode_literals. They make reasoning about code more difficult and causes mysterious failures when a function is moved from one module to another.

**Comment by wardi** [comment]
This is a really trivial issue.

4/5 of the `ur''` strings contain no backslashes so we can drop the `r` without any ill effect. The last one we can just double up the backslash in a normal `u''` string. That way we're not relying on the python 2 automatic conversion to unicode and we don't have to make a special case in our coding standards tests.

**Comment by cclauss** [comment]
That is what #4052 does.
